content/2020-04-23-value-range-analysis.rst

```diff
@@ -143,13 +196,46 @@ then prints the python code back, gives:
         return (a_, b, 10)
 
 The two first prints have been removed, because they were guarded by conditions
-that never hold. ``b == 9`` never holds because ``b`` starts at ``10`` and
-only grows. ``b == 10`` never holds either because the while loop is always
-entered at least once, and Pythran knows that thanks to the ``assert``.
+that never hold.
+
+Programming Nits
+================
+
+Using a naive control-flow based approach has some advantages. For instance, in the
+following code:
+
+.. code-block:: python
+
+    def foo(a):
+        if a > 12:
+            b = 1
+        else:
+            b = 2
+        if a > 1:
+            b = 2
+        return b
+
+Because the analysis explores the control flow graph in a depth-first manner,
+when visiting the children of ``if a > 12``, it finds ``if a > 1`` and knows
+*for sure* that the condition holds, thus ending up with ``b == 2`` upon the
+return. Then when visiting the ``else`` branch, it records ``b == 2`` and also
+ends up with ``b == 2`` upon the return.
+
+In the end, ``pythran -P`` on the above snippets yields:
+
+.. code-block:: python
+
+    def foo(a):
+        return 2
+
 
-And as Pythran doesn't perform any kind of modelisation of the accumulation in
-``a`` and ``b``, it cannot get smarter than this, but well, that's already a
-decent start ;-)
+Let's be honest, this algorithm is super greedy, and if we find a sequence of ``if``
+statements, it has an exponential complexity (and this happens a soon as we
+unroll a loop with a condition in its body). In that case we fall back to a
```

Comment:
Is there a number-of-if's threshold where we switch? May be nice to mention here.

Reply:
Also maybe link to the file where this analysis happens? Although I'm not too sure about that, so this is mostly open.